### PR TITLE
📝 add instructions for setting up vscode IDE

### PR DESCRIPTION
**Pull request by Sycha**
* [ ] Your changes are not possible to do through a plugin and relevant
    to a large audience (ideally all users of OctoPrint)
  * [ ] If your changes are large or otherwise disruptive: You have
    made sure your changes don't interfere with current development by
    talking it through with the maintainers, e.g. through a
    Brainstorming ticket
  * [ ] Your PR targets OctoPrint's devel branch if it's a completely
    new feature, or maintenance if it's a bug fix or improvement of
    existing functionality for the current stable version (no PRs
    against master or anything else please)
  * [ ] Your PR was opened from a custom branch on your repository
    (no PRs from your version of master, maintenance or devel please),
    e.g. dev/my_new_feature or fix/my_bugfix
  * [ ] Your PR only contains relevant changes: no unrelated files,
    no dead code, ideally only one commit - rebase and squash your PR
    if necessary!
  * [ ] Your changes follow the existing coding style
  * [ ] If your changes include style sheets: You have modified the
    .less source files, not the .css files (those are generated with
    lessc)
  * [ ] You have tested your changes (please state how!) - ideally you
    have added unit tests
  * [ ] You have run the existing unit tests against your changes and
    nothing broke
  * [ ] You have added yourself to the AUTHORS.md file :)

#### What does this PR do and why is it necessary?
Adds vscode IDE setup instructions to the documentation, necessary as vscode is a popular IDE and therefore likely to be used by potential contributors.

#### How was it tested? How can it be tested by the reviewer?
Has been tested manually in an Ubuntu 20.04 VM. Install vscode and run through the instructions

### CODE REVIEW

File: docs/development/environment.rst [hunk 1]
```diff
@@ -387,3 +387,101 @@ Otherwise go through Settings.
    Make sure you are running a PyCharm version of 2016.1 or later, or manually fix
    `a debugger bug contained in earlier versions <https://youtrack.jetbrains.com/issue/PY-18365>`_ or plugin management
    will not work in your developer install when running OctoPrint from PyCharm in debug mode.
+
+Visual Studio Code (vscode)
+---------------------------
+  - Install Visual Studio from https://code.visualstudio.com/Download
+  - Open folder select OctoPrint checkout folder (e.g. ``~/devel/OctoPrint`` or ``C:\Devel\OctoPrint``)
+    settings.json
```
Comment (cp2004) [inline docs/development/environment.rst:395]: Doesn't make sense here
```suggestion
```

File: docs/development/environment.rst [hunk 2]
```diff
@@ -387,3 +387,101 @@ Otherwise go through Settings.
    Make sure you are running a PyCharm version of 2016.1 or later, or manually fix
    `a debugger bug contained in earlier versions <https://youtrack.jetbrains.com/issue/PY-18365>`_ or plugin management
    will not work in your developer install when running OctoPrint from PyCharm in debug mode.
+
+Visual Studio Code (vscode)
+---------------------------
+  - Install Visual Studio from https://code.visualstudio.com/Download
```
Comment (cp2004) [inline docs/development/environment.rst:393]:
```suggestion
  - Install Visual Studio Code from https://code.visualstudio.com/Download
```

File: docs/development/environment.rst [hunk 3]
```diff
@@ -387,3 +387,101 @@ Otherwise go through Settings.
...
+
+
+    launch.json
+      .. code-block:: json
+
+        {
+          "version": "0.2.0",
+          "configurations": [
+              {
+                  "name": "Octoprint",
```
Comment (cp2004) [inline docs/development/environment.rst:452]:
```suggestion
                  "name": "OctoPrint",
```
Very picky this one... 🙂

File: docs/development/environment.rst [hunk 4]
```diff
@@ -387,3 +387,101 @@ Otherwise go through Settings.
...
+  In vscode terminal, or with venv active install code formatter black and linter flake8 by running:
+    .. code-block:: bash
+
+      python -m pip install -U black flake8 flake8-bugbear
+
+  Summary of vscode config:
+
+  * Pressing f5 will now start OctoPrint in debug
+
+  * Your terminal inside vscode uses the virtrual python environment
```
Comment (cp2004) [inline docs/development/environment.rst:480]:
```suggestion
  * Your terminal inside vscode uses the virtual python environment
```

**Review by cp2004 — CHANGES_REQUESTED** [review]
Very small (& picky) typos/word changes

**Comment by Sycha** [comment]
Thanks @cp2004 for catching those typos. All good :+1: 

First time using github pull request review feature do let me know if I need to do something differently. Cheers

File: docs/development/environment.rst [hunk 5]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
    Make sure you are running a PyCharm version of 2016.1 or later, or manually fix
    `a debugger bug contained in earlier versions <https://youtrack.jetbrains.com/issue/PY-18365>`_ or plugin management
    will not work in your developer install when running OctoPrint from PyCharm in debug mode.
+
+Visual Studio Code (vscode)
+---------------------------
+  - Install Visual Studio Code from https://code.visualstudio.com/Download
```
Comment (foosel) [inline docs/development/environment.rst:393]: Somehow the missing newline here irks me 😅

```suggestion

  - Install Visual Studio Code from `code.visualstudio.com <https://code.visualstudio.com/Download>`_
```

File: docs/development/environment.rst [hunk 6]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+  - Open folder select OctoPrint checkout folder (e.g. ``~/devel/OctoPrint`` or ``C:\Devel\OctoPrint``)
+
+  - Create a directory '.vscode' if not already present in the root of the project
```
Comment (foosel) [inline docs/development/environment.rst:396]:
```suggestion
  - Create a directory ``.vscode`` if not already present in the root of the project
```

File: docs/development/environment.rst [hunk 7]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+
+  - Create the following files inside the .vscode directory
```
Comment (foosel) [inline docs/development/environment.rst:398]:
```suggestion
  - Create the following files inside the ``.vscode`` directory
```

File: docs/development/environment.rst [hunk 8]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+                      "--debug"
+                  ],
+                  "cwd": "${workspaceFolder}/src",
+                  "preLaunchTask": "clean build artifacts"
+              }
+          ]
+        }
+  
+  In the terminal install the python extension by running this command:
+    .. code-block:: bash
```
Comment (foosel) [inline docs/development/environment.rst:466]: Newline

```suggestion

    .. code-block:: bash
```

File: docs/development/environment.rst [hunk 9]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+
+      code --install-extension ms-python.python
+
+  In vscode terminal, or with venv active install code formatter black and linter flake8 by running:
+    .. code-block:: bash
```
Comment (foosel) [inline docs/development/environment.rst:471]: Newline

```suggestion

    .. code-block:: bash
```

File: docs/development/environment.rst [hunk 10]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+
+      python -m pip install -U black flake8 flake8-bugbear
+
+  Summary of vscode config:
+
+  * Pressing f5 will now start OctoPrint in debug
```
Comment (foosel) [inline docs/development/environment.rst:477]:
```suggestion
  * Pressing ``F5`` will now start OctoPrint in debug mode
```

File: docs/development/environment.rst [hunk 11]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+
+  * Your terminal inside vscode uses the virtual python environment
+
+  * Saving a file will run an auto formatter and import sort
+  
+  * Ctrl + shift + b can be used to run the 'build docs' task to rebuild the documentation
```
Comment (foosel) [inline docs/development/environment.rst:483]:
```suggestion
  * ``Ctrl+Shift+B`` can be used to run the ``build docs`` task to rebuild the documentation
```

File: docs/development/environment.rst [hunk 12]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+            ],
+            "[python]": {
+                "editor.codeActionsOnSave": {
+                    "source.organizeImports": true
+                }
+            },
+            "python.linting.pylintEnabled": false,
+            "python.linting.flake8Enabled": true,
+            "python.linting.enabled": true
+        }
```
Comment (foosel) [inline docs/development/environment.rst:422]: Something's not right with that indent...

```suggestion
      .. code-block:: json

         {
             "python.defaultInterpreterPath": "venv3/bin/python",
             "python.formatting.provider": "black",
             "python.formatting.blackArgs": [
                 "--config",
                 "black.toml"
             ],
             "editor.formatOnSave": true,
             "python.sortImports.args": [
                 "--profile=black",
             ],
             "[python]": {
                 "editor.codeActionsOnSave": {
                     "source.organizeImports": true
                 }
             },
             "python.linting.pylintEnabled": false,
             "python.linting.flake8Enabled": true,
             "python.linting.enabled": true
         }
```

File: docs/development/environment.rst [hunk 13]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+                "type": "shell",
+                "command": "python ./setup.py clean"
+            },
+            {
+                "label": "build docs",
+                "type": "shell",
+                "command": "sphinx-build -b html ./docs ./docs/_build"
+            }
+          ]
+        }
```
Comment (foosel) [inline docs/development/environment.rst:441]: Indent

```suggestion
      .. code-block:: json

         {
           "version": "2.0.0",
           "tasks": [
             {
                 "label": "clean build artifacts",
                 "type": "shell",
                 "command": "python ./setup.py clean"
             },
             {
                 "label": "build docs",
                 "type": "shell",
                 "command": "sphinx-build -b html ./docs ./docs/_build"
             }
           ]
         }
```

File: docs/development/environment.rst [hunk 14]
```diff
@@ -387,3 +387,99 @@ Otherwise go through Settings.
...
+                  "module": "octoprint",
+                  "args": [
+                      "serve",
+                      "--debug"
+                  ],
+                  "cwd": "${workspaceFolder}/src",
+                  "preLaunchTask": "clean build artifacts"
+              }
+          ]
+        }
```
Comment (foosel) [inline docs/development/environment.rst:463]: Indent

```suggestion
      .. code-block:: json

         {
           "version": "0.2.0",
           "configurations": [
               {
                   "name": "OctoPrint",
                   "type": "python",
                   "request": "launch",
                   "module": "octoprint",
                   "args": [
                       "serve",
                       "--debug"
                   ],
                   "cwd": "${workspaceFolder}/src",
                   "preLaunchTask": "clean build artifacts"
               }
           ]
         }
```

**Review by foosel — CHANGES_REQUESTED** [review]
Also found some minor issues (formatting primarily). Other than that it looks good :)

I'm wondering if there's a *built-in* option to use pre-commit for black, isort and flake8. As mentioned on Signal I currently use the Run on Save extension to do that, but of course that requires an additional extension whereas the currently documented approach here doesn't. 

Might make sense to maybe add black, isort and flake8 to the devel dependencies to get rid of one manual install step here if there's no built-in possibility?

**Review by foosel — APPROVED** [review]
Thumbs up from my side, not sure if @cp2004 still sees open points.

**Review by cp2004 — APPROVED** [review]

**Comment by Sycha** [comment]
Adding black, flake8, isort to the devel dependencies seems to make sense to me, though having two different environments (pre-commit/devel) try and run the same task does introduce the risk of them diverging and giving inconsistent behaviour, so not the perfect solution.

I haven't been able to find a built in way to run the pre-commit on save so for now went with the vscode settings to run it from the built in formatter settings which happen to use the virtual environment.

This [extension](https://marketplace.visualstudio.com/items?itemName=jkillian.custom-local-formatters) suggests that adding new code formatters can only be done through an extension.

So currently either you get a consistent build environment by using an extension to use pre-commit for formatting but rely on ANOTHER vscode extension which could change at a moments notice or you get a virtual environment dependant formatter which could be a pain when working with legacy code and could be using different package versions to pre-commit.

Bit of a judgement call, but I am leaning towards a consistent use of pre-commit even though I dislike having another extension installed.

**Comment by thelastWallE** [comment]
Nice addition for me as i use vsc!

There is also something i wanted to point out that is not related to this commit but is related to this instructions file. In the instructions on the first "half" of the instructions in environment.rst is like a copy of the second part where the instructions for windows and linux are separate.
I was using the first half to install octoprint in the venv, but after this i was reading the second part of the file and it was better to understand for me. I think the first part needs some cleaning of the things that are when again written in the second part.

I would do this myself, but i give the warning that i'm not a professional programmer and also not native english speaking.
Some last thing. If i would like to switch from python 2 to 3 and versa in vsc. Would i need to have like two setting.json files? Because of the ` "python.defaultInterpreterPath": "venv3/bin/python",`

**Pull request merged**